server/src/main/java/com/hedera/block/server/persistence/storage/read/BlockAsLocalFileReader.java

```diff
@@ -17,34 +17,65 @@
 package com.hedera.block.server.persistence.storage.read;
 
 import com.hedera.block.common.utils.Preconditions;
+import com.hedera.block.server.persistence.storage.path.BlockPathResolver;
 import com.hedera.hapi.block.BlockUnparsed;
+import com.hedera.pbj.runtime.ParseException;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
 import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Objects;
 import java.util.Optional;
 
 /**
  * A Block reader that reads block-as-file.
  */
 public final class BlockAsLocalFileReader implements LocalBlockReader<BlockUnparsed> {
+    private final BlockPathResolver pathResolver;
+
     /**
      * Constructor.
+     *
+     * @param pathResolver valid, {@code non-null} instance of
+     * {@link BlockPathResolver} used to resolve paths to block files
      */
-    private BlockAsLocalFileReader() {}
+    private BlockAsLocalFileReader(@NonNull final BlockPathResolver pathResolver) {
+        this.pathResolver = Objects.requireNonNull(pathResolver);
+    }
 
     /**
      * This method creates and returns a new instance of
      * {@link BlockAsLocalFileReader}.
      *
+     * @param pathResolver valid, {@code non-null} instance of
+     * {@link BlockPathResolver} used to resolve paths to block files
      * @return a new, fully initialized instance of
      * {@link BlockAsLocalFileReader}
      */
-    public static BlockAsLocalFileReader newInstance() {
-        return new BlockAsLocalFileReader();
+    public static BlockAsLocalFileReader of(@NonNull final BlockPathResolver pathResolver) {
+        return new BlockAsLocalFileReader(pathResolver);
     }
 
     @NonNull
     @Override
-    public Optional<BlockUnparsed> read(final long blockNumber) {
+    public Optional<BlockUnparsed> read(final long blockNumber) throws IOException, ParseException {
         Preconditions.requireWhole(blockNumber);
-        throw new UnsupportedOperationException("Not implemented yet");
+        final Path resolvedBlockPath = pathResolver.resolvePathToBlock(blockNumber);
+        if (Files.exists(resolvedBlockPath)) {
+            return Optional.of(doRead(resolvedBlockPath));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private BlockUnparsed doRead(final Path resolvedBlockPath) throws IOException, ParseException {
+        // todo there are other ways of parsing, for example we can pass ReadableSequentialData instead of the fos
+        // because the readAllBytes does not seem like the best idea, what is the correct (or better) way to parse
+        // the block file?
```

Comment:
Opinions?

Reply:
I am assuming, here, that you intend to just wrap the input stream in the readable sequential data and pass that to the parse method instead of calling `readAllBytes`.

I believe that constructing a `ReadableSequentialData` here, instead of reading the full file as bytes before parsing, will save a byte array copy, and should improve memory efficiency for minimal change in complexity.

Reply:
✅ Indeed, wrapping the input stream is what I meant. I've addressed that, thanks!